drivers/serial/uart_cdns.c

```diff
@@ -0,0 +1,305 @@
+
+	uart_regs->rx_timeout = DEFAULT_RTO_PERIODS_FACTOR;
+	uart_regs->intr_enable |= (CSR_RTRIG_MASK | CSR_RBRK_MASK | CSR_TOUT_MASK);
+}
+
+/** Disable RX UART interrupt */
+void uart_cdns_disable_rx_irq(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->intr_disable |= (CSR_RTRIG_MASK | CSR_RBRK_MASK | CSR_TOUT_MASK);
+}
+
+static int uart_cdns_irq_rx_ready(const struct device *dev)
+{
+	return !uart_cdns_is_rx_fifo_empty(DEV_UART(dev));
+}
+
+static void uart_cdns_enable_irq_err(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->intr_enable |=
+		(CSR_TOVR_MASK | CSR_TOUT_MASK | CSR_PARE_MASK | CSR_FRAME_MASK | CSR_ROVR_MASK);
+}
+
+static void uart_cdns_disable_irq_err(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	uart_regs->intr_disable |=
+		(CSR_TOVR_MASK | CSR_TOUT_MASK | CSR_PARE_MASK | CSR_FRAME_MASK | CSR_ROVR_MASK);
+}
+
+static int uart_cdns_is_irq_pending(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+
+	return (uart_regs->channel_intr_status != 0);
+}
+
+/** Check for IRQ updates */
+static int uart_cdns_update_irq(const struct device *dev)
+{
+	return 1;
+}
+
+/** Set the callback function pointer for IRQ. */
+void uart_cdns_set_irq_callback(const struct device *dev, uart_irq_callback_user_data_t cb,
+				   void *cb_data)
+{
+	struct uart_cdns_data *data;
+
+	data = DEV_DATA(dev);
+	data->callback = cb;
+	data->cb_data = cb_data;
+}
+
+static void uart_cdns_irq_handler(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+	uint32_t key = irq_lock();
+	uint32_t isr_status;
+	struct uart_cdns_data *data = DEV_DATA(dev);
+
+	if (data->callback) {
+		data->callback(dev, data->cb_data);
+	}
+
+	/* clear events */
+	/* need to make local copy of the status */
+	isr_status = uart_regs->channel_intr_status;
+
+	irq_unlock(key);
+}
+
+#endif
+
+static const struct uart_driver_api uart_cdns_driver_api = {
+	.poll_in = uart_cdns_poll_in,
+	.poll_out = uart_cdns_poll_out,
+	.err_check = NULL,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = uart_cdns_fill_fifo,
+	.fifo_read = uart_cdns_read_fifo,
+	.irq_tx_enable = uart_cdns_enable_tx_irq,
+	.irq_tx_disable = uart_cdns_disable_tx_irq,
+	.irq_tx_ready = uart_cdns_irq_tx_ready,
+	.irq_rx_enable = uart_cdns_enable_rx_irq,
+	.irq_rx_disable = uart_cdns_disable_rx_irq,
+	.irq_rx_ready = uart_cdns_irq_rx_ready,
+	.irq_err_enable = uart_cdns_enable_irq_err,
+	.irq_err_disable = uart_cdns_disable_irq_err,
+	.irq_is_pending = uart_cdns_is_irq_pending,
+	.irq_update = uart_cdns_update_irq,
+	.irq_callback_set = uart_cdns_set_irq_callback
+#endif
+};
+
+/** Initialize the UART */
+static int uart_cdns_init(const struct device *dev)
+{
+	struct uart_cdns_regs *uart_regs = DEV_UART(dev);
+	const struct uart_cdns_device_config *const dev_cfg = DEV_CFG(dev);
+
+	if (!dev_cfg->do_not_init) {
```

Comment:
Why would you need a `do_not_init` here for this IP? I though it should be expected that after the init the IP is ready to go. Without any of these writes, I would expect it to not be

Reply:
do_not_init variable gets its value from the DTS property, which if not present, will default to false, and initialize the UART IP. This was added as an additional feature for cases if a MCU wants to access the IP, but not reinitialize it if previously initialized.
 
However, if it seems to create more confusion, i can remove it and make it simpler. Please let me know your thoughts.

Reply:
Updated. Thanks.